Implemented things by branch name, also:

b0r1ngx opened the pull request:
1) Created setContentDescription modifier that allows to set contentDescription on any used Modifier - as accessibility feature, suppossing that it may help to understand the UI architecture components on the screen more

Not tested how it would work for mines >999, or <-99, things that code logic is draw more digits, how much it needs

b0r1ngx committed:
Implemented things by branch name, also:
1) Created setContentDescription modifier that allows to set contentDescription on any used Modifier - as accessibility feature, suppossing that it may help to understand the UI architecture more

b0r1ngx commented:
To understand how it looks like, attach few screenshots:
When start a game:
![image](https://github.com/arkivanov/Minesweeper/assets/55315908/27b934f1-d6a4-4bdb-9f14-6df3c3eccf5c)

when user mark more than max of the mines:
![image](https://github.com/arkivanov/Minesweeper/assets/55315908/4bf109ae-96d0-4277-b574-396162b00d16)

b0r1ngx committed:
"bomb/s" to "mine/s" to be more project-convinient

b0r1ngx commented:
@arkivanov Hello, please, check how it looks like!

arkivanov commented on composeApp/src/commonMain/kotlin/com/arkivanov/minesweeper/game/GameState.kt:
```diff
@@ -98,5 +98,6 @@ internal val CellStatus.isFlagged: Boolean
 internal fun Cell.open(): Cell =
     copy(status = CellStatus.Open)
 
+// TODO: Can we somehow increase work of this thing? not each time count run for all grid
```
Let's change the TODO text to something like "Consider making this a part of the state"

arkivanov commented on composeApp/src/commonMain/kotlin/com/arkivanov/minesweeper/ComposeUtils.kt:
```diff
@@ -0,0 +1,15 @@
+import androidx.compose.ui.semantics.role
+import androidx.compose.ui.semantics.semantics
+
+fun Modifier.setContentDescription(description: String, role: Role): Modifier =
```
Let's name it as `semantics`.

arkivanov commented on composeApp/src/commonMain/kotlin/com/arkivanov/minesweeper/game/GameContent.kt:
```diff
@@ -126,6 +131,43 @@ internal fun GameContent(component: GameComponent, modifier: Modifier = Modifier
 }
 
+@Composable
+fun RemainingMines(
```
I think we can leverage the `padStart` function and do something as follows. This `Counter` function could be reused in both cases - remaining mine counter and (future) time counter.

```kotlin
@Composable
private fun Counter(value: Int, modifier: Modifier = Modifier) {
    Row(modifier = modifier, horizontalArrangement = Arrangement.Center) {
        value.toCounterString().forEach { char ->
            Image(
                modifier = Modifier.size(width = 13.dp, height = 23.dp),
                painter = LocalGameIcons.icons.digits.getValue(char),
                contentDescription = char.toString(),
            )
        }
    }
}

private fun Int.toCounterString(): String =
    if (this >= 0) {
        coerceAtMost(999).toString().padStart(length = COUNTER_LENGTH, padChar = '0')
    } else {
        "-" + coerceAtLeast(-99).absoluteValue.toString().padStart(length = COUNTER_LENGTH - 1, padChar = '0')
    }
```

arkivanov commented on composeApp/src/commonMain/kotlin/com/arkivanov/minesweeper/game/GameContent.kt:
```diff
@@ -126,6 +131,43 @@ internal fun GameContent(component: GameComponent, modifier: Modifier = Modifier
+                    painter = LocalGameIcons.icons.digits.getValue('0'),
+                    contentDescription = "0",
+                    contentScale = ContentScale.Crop,
+                    modifier = Modifier.height(26.dp),
```
Let's define the exact size for the image as per PNGs: 13.dp x 23.dp

arkivanov commented on composeApp/src/commonMain/kotlin/com/arkivanov/minesweeper/game/GameContent.kt:
```diff
@@ -52,8 +52,11 @@ import androidx.compose.ui.unit.dp
+import com.arkivanov.minesweeper.setContentDescription
 
 private val cellSize = 16.dp
+private const val defaultCountersSize = 3
```
I suggest to name it as `COUNTER_LENGTH`.

arkivanov commented on composeApp/src/commonMain/kotlin/com/arkivanov/minesweeper/ComposeUtils.kt:
```diff
@@ -0,0 +1,15 @@
+import androidx.compose.ui.semantics.semantics
+
+fun Modifier.setContentDescription(description: String, role: Role): Modifier =
+    this.then(
```
Can we just call `semantics { ... }` here?

b0r1ngx committed:
Changes-recommendation, by reviewer:
1) decrease size of setSemantics, also rename in from setContentDescription

2) deprecate RemainingMines, create Counter, toCounterString

3) changes TODO idea

b0r1ngx committed:
delete not used things

b0r1ngx committed:
using setSemantics

b0r1ngx commented:
@arkivanov Hello, I'm made recommended changes, please, check how it looks like!

arkivanov commented on composeApp/src/commonMain/kotlin/com/arkivanov/minesweeper/game/GameContent.kt:
```diff
@@ -72,11 +74,12 @@ internal fun GameContent(component: GameComponent, modifier: Modifier = Modifier
-                        textAlign = TextAlign.Center
+                    Counter(
+                        value = remainingMines,
+                        modifier = Modifier.weight(1f).setSemantics(
```
Now after the changes, maybe we can just use the original `semantics` modifier? What benefits the custom function gives us?

arkivanov commented on composeApp/src/commonMain/kotlin/com/arkivanov/minesweeper/game/GameContent.kt:
```diff
@@ -86,12 +89,7 @@ internal fun GameContent(component: GameComponent, modifier: Modifier = Modifier
-                        modifier = Modifier.weight(1f),
-                        textAlign = TextAlign.Center
-                    )
+                    Stopwatch(modifier = Modifier.weight(1f))
```
Let's just use `Counter(value = 0, modifier = Modfier.weight(1F))` here for now.

b0r1ngx committed:
1) using semantics, instead same setSemantics

2) Create TODO

b0r1ngx commented:
Let's rock it!

arkivanov commented:
Thank you!

b0r1ngx commented:
> Thank you!

Thank you too, for reviewing my code and making it better together! 

Hope I will start implementing the stopwatch soon, I don't want to use additional dependencies (for educational purposes for me and project, if someone wants to research it)

Also I tried to go to wasm site from this repo, I also got amazed that you create a pipeline that automatically creates a build and publishes it to the site (you did a great work with this CI/CD)

I visit it from mobile android chrome browser, and can't click anything. is this wasm restriction for now? Or it just doesn't work because in the project clicks are used instead of touches?

arkivanov commented:
Yeah, Wasm on Android chrome is pretty much broken currently. Also on any browser, all mouse clicks are recognised as left clicks.
